File: chroma-league/src/main/java/net/booone/chromaleague/hud/parts/resource/RumbleHeatBar.java
package net.booone.chromaleague.hud.parts.resource;

import net.booone.chromaleague.hud.parts.ProgressBar;
import net.booone.chromaleague.state.GameStateHelper;
import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.Frame;
import net.booone.razersdk.color.Color;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.color.TransitionColor;

import java.time.Duration;
import java.time.LocalTime;

public class RumbleHeatBar extends AnimatedFrame {
    private static final double HEAT_THRESHOLD = 8.0;
    private static final TransitionColor HEAT_COLOR = new TransitionColor(StaticColor.YELLOW, StaticColor.RED);

    private LocalTime lastHeatingTime = LocalTime.now();
    private int previousResourcePercentage;
    private int highestResourceValue;
    private boolean overheating;
    private boolean coolingDown;

    @Override
    public Frame getFrame() {
        final int resourcePercentage = GameStateHelper.getResourcePercentage();
        if (resourcePercentage != previousResourcePercentage) {
            final LocalTime now = LocalTime.now();
            if (resourcePercentage < previousResourcePercentage) {
                if (!coolingDown) {
                    highestResourceValue = previousResourcePercentage;
                }
                coolingDown = true;
                final long durationOfChange = Duration.between(lastHeatingTime, now).toMillis();
                final int delta = highestResourceValue - resourcePercentage;
                double decayRate = delta * 1000.0 / durationOfChange;
                overheating = decayRate > HEAT_THRESHOLD;
            }
            if (resourcePercentage > previousResourcePercentage) {
                coolingDown = false;
                overheating = false;
                lastHeatingTime = now;
                highestResourceValue = resourcePercentage;
            }
            previousResourcePercentage = resourcePercentage;
        }

        Color color;
        if (overheating) {
            color = StaticColor.RED;
        } else {
            if (resourcePercentage >= 50) {
                color = HEAT_COLOR.getColorAtPercent(resourcePercentage - 50 << 1);
            } else {
                color = StaticColor.WHITE;
            }
        }

        addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), resourcePercentage, color));
        return super.getFrame();
    }
}

File: src/hud/parts/resource/RumbleHeatBar.ts
import {AnimatedFrame} from '../../../razer-sdk/animation/AnimatedFrame.js';
import {Frame} from '../../../razer-sdk/animation/Frame.js';
import {Color} from '../../../razer-sdk/color/Color.js';
import {StaticColor} from '../../../razer-sdk/color/StaticColor.js';
import {TransitionColor} from '../../../razer-sdk/color/TransitionColor.js';
import {ProgressBar} from '../ProgressBar.js';
import {getResourceBarKeys} from './ResourceBars.js';
import {getResourcePercentage} from '../../../state/GameStateHelper.js';

const HEAT_THRESHOLD = 8.0;
const HEAT_COLOR = new TransitionColor(StaticColor.YELLOW, StaticColor.RED);

export class RumbleHeatBar extends AnimatedFrame {
    private lastHeatingTime = Date.now();
    private previousResourcePercentage = 0;
    private highestResourceValue = 0;
    private overheating = false;
    private coolingDown = false;

    pollFrame(): Frame {
        const resourcePercentage = getResourcePercentage();
        if (resourcePercentage !== this.previousResourcePercentage) {
            const now = Date.now();
            if (resourcePercentage < this.previousResourcePercentage) {
                if (!this.coolingDown) {
                    this.highestResourceValue = this.previousResourcePercentage;
                }
                this.coolingDown = true;
                const durationOfChange = now - this.lastHeatingTime;
                const delta = this.highestResourceValue - resourcePercentage;
                const decayRate = delta * 1000.0 / durationOfChange;
                this.overheating = decayRate > HEAT_THRESHOLD;
            }
            if (resourcePercentage > this.previousResourcePercentage) {
                this.coolingDown = false;
                this.overheating = false;
                this.lastHeatingTime = now;
                this.highestResourceValue = resourcePercentage;
            }
            this.previousResourcePercentage = resourcePercentage;
        }

        let color: Color;
        if (this.overheating) {
            color = StaticColor.RED;
        } else {
            if (resourcePercentage >= 50) {
                color = HEAT_COLOR.getColorAtPercent((resourcePercentage - 50) << 1);
            } else {
                color = StaticColor.WHITE;
            }
        }

        this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), resourcePercentage, color));
        return super.pollFrame();
    }
}
